### ISSUE TITLE
Invalid Files get parsed anyway

### ISSUE THREAD

**Issue by dhofstetter** [tefra/xsdata#560]
Hello,
I really love this package, it offers a very modern way to handle xml files with their schemas.

At the moment I'm struggling a little bit, because I can parse files that are wrong to the schema spec for me.
I'm not sure if this is expected behaviour. I prepared an example to show my problem (appended as zip file).

[example.zip](https://github.com/tefra/xsdata/files/6855758/example.zip)

The schema used for generating the classes: 

```
<?xml version="1.1" encoding="UTF-8"?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:rap="urn:example.com:some:document:1:0" targetNamespace="urn:example.com:some:document:1:0">
	<xs:element name="SomeDocument">
		<xs:complexType>
			<xs:sequence>
				<xs:element ref="rap:SenderRole" minOccurs="1" maxOccurs="1"/>
				<xs:element ref="rap:ReceiverRole" minOccurs="1" maxOccurs="1"/>
			</xs:sequence>
		</xs:complexType>
	</xs:element>
	<xs:element name="SenderRole">
		<xs:complexType>
			<xs:simpleContent>
				<xs:extension base="xs:string">
					<xs:attribute name="v" type="xs:string" use="required" fixed="A34"/>
				</xs:extension>
			</xs:simpleContent>
		</xs:complexType>
	</xs:element>
	<xs:element name="ReceiverRole">
		<xs:complexType>
			<xs:simpleContent>
				<xs:extension base="xs:string">
					<xs:attribute name="v" type="xs:string" use="required" fixed="A27"/>
				</xs:extension>
			</xs:simpleContent>
		</xs:complexType>
	</xs:element>
</xs:schema>
```

My problem is that even with minOccurs="1" that the field is generated as:

```
@dataclass
class ReceiverRole:
    class Meta:
        namespace = "urn:example.com:some:document:1:0"

    value: Optional[str] = field(
        default=None,
        metadata={
            "required": True,
        }
    )
    v: str = field(
        init=False,
        default="A27",
        metadata={
            "type": "Attribute",
        }
    )
```

Within my example I commented within the lines where the behaviour is not consistent to me. Whenever I parse an xml like this: 

```
<?xml version="1.0" encoding="UTF-8"?>
<SomeDocument xsi:schemaLocation="urn:example.com:some:document:1:0 ./urn-some-document.xsd" xmlns="urn:example.com:some:document:1:0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
    <SenderRole v="A34"/>
    <ReceiverRole v="A34"/>
</SomeDocument>
```
It works, with only that the required fields are None if the value is missing within the xml and is the constant value if a wrong value is within the xml.

Can you explain me, if this is expected behaviour and why? Do you have a suggestion if I want to get errors when parsing the file?.

BR Daniel

**Comment by tefra** [comment]
Hi @dhofstetter there are a couple of issues in your examples

first of all xsdata really suffers because of [this](https://docs.python.org/3.10/library/dataclasses.html),

> TypeError will be raised if a field without a default value follows a field with a default value. This is true whether this occurs in a single class, or as a result of class inheritance.

To overcome this in 3.6 <= python <= 3.9 xsdata is turning all fields as optional if there is no default value!

Hopefully in python 3.10 there is a new directive `kw_only` that will take care of that (wip #556), I am also working on an [attrs](https://www.attrs.org/) plugin that already offers this feature. This will allow the parser to fail if a required property is not initialized :rocket: 

But in general, yes xsdata is not really an xsd validator. The input documents are supposed to be valid against the schema and that is the user's responsibility. At least that's the most common behavior in binding libraries. Despite what I just said  :smile:  the parser will fail on unknown properties or converter warnings based on the [config](https://xsdata.readthedocs.io/en/latest/api/reference/xsdata.formats.dataclass.parsers.config.ParserConfig.html#xsdata.formats.dataclass.parsers.config.ParserConfig) as a security measure against possible attacks.

Maybe we could also add as a config property to verify that fixed values if present in the document match the model definition.

**Comment by dhofstetter** [comment]
hmm I thought binding and validation goes hand in hand (thought it's supposed to be a bit like pydantic just for xsd/xml stuff instead of json).

So probably for my use case - as I cannot garantuee that my inupts are valid all the time - I will have to use another library to match xsd against the xml and parse it with xsdata afterwards (being sure that valid documents get mapped as expected).

The typeerror mentioned; does it occur for faulty documents only or in daily business as well? So is it that initializing a dataclass won't work if you provide the init values in the wrong order?

**Comment by tefra** [comment]
That's a quote from the official python dataclasses documentation,

In python < 3.10 this will fail

```python
@dataclass
class Foo:
    a: int = 4
    b: str
```

that's why xsdata marks all fields without default values as optional

```python
@dataclass
class Foo:
    a: int = 4
    b: Optional[str] = None
```

But in  3.10 this will work, making required validation work out of the box, aka raising exception if the required field is not initialized. The order you pass the params to the constructor won't matter as all of them will be keyword arguments

```python
@dataclass(kw_only=True)
class Foo:
    a: int = 4
    b: str
```

**Comment by dhofstetter** [comment]
hmm I see. Is it possible to sort the fields after collecting some sort mapping containing all values? So before generating the data classes it might be possible to sort them by being optional and default in the beginning and afterwards the required fields. Or do I miss something?

**Comment by tefra** [comment]
xml elements ordering is important, kw_only generator config was just merged for python 3.10, for versions <= 3.9 you will have to wait for the attrs plugin
